packages/components/src/utils/graph.ts

```diff
@@ -0,0 +1,150 @@
+import type { InternalNode } from '@xyflow/react';
+import { Position } from '@xyflow/react';
+
+type Params = [number, number, Position];
+type IntersectionPoint = { x: number; y: number };
+type EdgeParams = {
+  sx: number;
+  sy: number;
+  tx: number;
+  ty: number;
+  sourcePos: Position;
+  targetPos: Position;
+};
+
+function getParams(nodeA: InternalNode, nodeB: InternalNode): Params {
+  const centerA = getNodeCenter(nodeA);
+  const centerB = getNodeCenter(nodeB);
+  const horizontalDiff = Math.abs(centerA.x - centerB.x);
+  const verticalDiff = Math.abs(centerA.y - centerB.y);
+
+  let position: Position;
+  if (horizontalDiff > verticalDiff) {
+    position = centerA.x > centerB.x ? Position.Left : Position.Right;
+  } else {
+    position = centerA.y > centerB.y ? Position.Top : Position.Bottom;
+  }
+
+  const [x, y] = getHandleCoordsByPosition(nodeA, position);
+  return [x, y, position];
+}
+
+function getHandleCoordsByPosition(node: InternalNode, handlePosition: Position): [number, number] {
+  const handle = node.internals.handleBounds?.source?.find(h => h.position === handlePosition);
+
+  if (!handle) {
+    throw new Error(`Handle not found for position: ${handlePosition}`);
+  }

```

Comment:
can be dangerous if you throw an error without catch it.
Mostly it is much better to just return undefined, because this will be shown by the typescript compiler so you are enforced to handle this case.
If you throw an error in a function no one will tell the user of this function that there should by a try catch around it.